wsrep_loader.c

```diff
@@ -224,8 +224,34 @@ void wsrep_unload(wsrep_t *hptr)
         if (hptr->free)
             hptr->free(hptr);
         if (hptr->dlh)
+        {
             dlclose(hptr->dlh);
+        }
         free(hptr);
     }
 }
 
+void wsrep_cleanup_unload(wsrep_t* hptr, void (*cleanup_fn)(void*),
+                          void* app_ctx)
+{
+    if (!hptr) {
+        logger (WSREP_LOG_WARN, "wsrep_unload(): null pointer.");
+    } else {
+        if (hptr->free)
+            hptr->free(hptr);
+        if (cleanup_fn)
+            (*cleanup_fn)(app_ctx);
```

Comment:
This method seems like an unnecessary complication of the API: it absolutely does not add any benefits or reduces boilerplate code. Application itself can cleanup its context any time it wishes after provider is unloaded.

Reply:
It is an additional free function, so not a big complication. The user can freely use `wsrep_unload()` if it is sufficient enough.

The reason for this function exist is to provide well defined deinitialization point for global data structures in provider library between freeing the provider object and unloading the library. Without this it would not be possible to reload the provider library cleanly.

Reply:
Turned out that the deinitialization can be done between `wsrep->free()` and `wsrep_unload()`, so the extra function can be removed.